### PR TITLE
Feat/label formatter

### PR DESCRIPTION
**Pull request by pairone**
[[中文版模板 / Chinese template](https://github.com/VisActor/VChart/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Release
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close https://github.com/VisActor/VChart/issues/1382

### 🐞 Bugserver case id
655b1dfb316a74008d2d40c0
655dc19812c3a41900817f35
655dc53512c3a41900817f37

### 💡 Background and solution
image
image

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vchart/__tests__/runtime/browser/test-page/bar.ts [hunk 1]
```diff
@@ -184,6 +184,10 @@ const run = () => {
           fill: '#1664FF'
         }
       }
+    },
+    label: {
+      visible: true,
+      formatter: 'catogory'
```
Comment (kkxxkk2019) [inline packages/vchart/__tests__/runtime/browser/test-page/bar.ts:190]: 这个 demo 不对吧，漏了 {}?

File: packages/vchart/package.json [hunk 2]
```diff
@@ -110,9 +110,9 @@
     "@size-limit/file": "9.0.0"
   },
   "dependencies": {
-    "@visactor/vutils": "~0.16.10",
-    "@visactor/vdataset": "~0.16.10",
-    "@visactor/vscale": "~0.16.10",
+    "@visactor/vutils": "0.16.15",
```
Comment (kkxxkk2019) [inline packages/vchart/package.json:113]: ~0.16.15

File: packages/vchart/src/component/label/interface.ts [hunk 3]
```diff
@@ -14,6 +14,13 @@ export interface ILabelSpec extends IComponentSpec {
   interactive?: boolean;
   /** 格式化函数 */
   formatMethod?: (text: string | string[], datum?: Datum, ctx?: ILabelFormatMethodContext) => string | string[];
+  /** 字符串模版
+   *  用{}包裹变量名的字符串模版, 变量名取自数据属性值
+   *  在饼图中支持配置百分比, {_percent_}
+   *  eg: 'type={type},value={value},percent={_percent_}'
+   *  since version '1.7.0'
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/label/interface.ts:21]: @since 1.7.0

File: packages/vchart/src/component/label/util.ts [hunk 4]
```diff
@@ -26,18 +27,31 @@ export enum LabelRule {
...
+export function textAttribute(
+  labelInfo: ILabelInfo,
+  datum: Datum,
+  formatMethod?: ILabelSpec['formatMethod'],
+  formatter?: ILabelSpec['formatter']
+) {
   const { labelMark, series } = labelInfo;
-  const field = series.getMeasureField()[0];
-  const textAttribute = { text: datum[field], data: datum } as any;
+  const textAttribute = { data: datum } as any;
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/label/util.ts:37]: 这里为啥把 text 属性去掉了？
image

File: packages/vchart/__tests__/runtime/browser/test-page/bar.ts (hunk 1 above)
Comment (kkxxkk2019) [reply packages/vchart/__tests__/runtime/browser/test-page/bar.ts:190]: 补充下 bugserver case

File: packages/vchart/src/component/label/util.ts [hunk 5]
```diff
@@ -26,18 +27,31 @@ export enum LabelRule {
...
   const attributes = Object.keys(labelMark.stateStyle.normal);
+
   for (const key of attributes) {
     const attr = labelMark.getAttribute(key as any, datum);
     textAttribute[key] = attr;
-    if (key === 'text' && formatMethod) {
-      textAttribute[key] = formatMethod(textAttribute[key], datum, { series });
+  }
+
+  if (formatMethod) {
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/label/util.ts:46]: 赞

File: packages/vchart/src/component/label/util.ts [hunk 6]
```diff
@@ -26,18 +27,31 @@ export enum LabelRule {
...
+    textAttribute.text = formatMethod(textAttribute.text, datum, { series });
+  }
+
+  if (formatter) {
+    if (series.type === 'pie') {
+      datum._percent_ = (datum[ARC_RATIO] * 100).toFixed(2) + '%';
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/label/util.ts:52]: 想起来了，对于百分比堆叠图，应该也是有占比的，可以判断下数据中有没有对应的字段，然后赋值给 datum._percent_

File: packages/vchart/__tests__/runtime/browser/test-page/bar.ts (hunk 1 above)
Comment (pairone) [reply packages/vchart/__tests__/runtime/browser/test-page/bar.ts:190]: 是的 这个是之前写的case了 忘记改了

**Review by kkxxkk2019 — APPROVED** [review]

**Pull request merged**